syntax = "proto3";  // 声明了protobuf的版本

package xgg;     // 声明了代码所在的包（对于C++来说是namespace）

// 定义下面的选项，表示生成service服务类和rpc方法描述，否则默认不生成
option cc_generic_services = true;

// 对于多个消息中重复的字段，可以打包成一个类型
message ResultCode
{
    int32 errcode = 1;
    bytes errmsg = 2;
}

// 数据 列表 映射表 一般就这交换的几种数据类型
// 定义登录请求消息类型 name pwd
message LoginRequest
{
    bytes name = 1;//代表第几个字段
    bytes pwd = 2;
}

// 定义登录响应消息类型
message LoginResponse
{
    ResultCode result = 1;
    bool success = 2;
}

message GetFriendListRequest
{
    uint32 userid = 1;
}

message User
{
    bytes name = 1;
    uint32 age = 2;
    enum Sex
    {
        MAN = 0;
        WOMAN = 1;
    }
    Sex sex = 3;
}

message GetFriendListResponse
{
    ResultCode result = 1;
    repeated User friend_list = 2;// 定义了一个<列表类型>
}

// ps：一般.proto里的stirng 最好用 bytes 代替 
// main.cpp的实际代码中和用字符串的效果一样

// 在protobuf里面怎么定义描述rpc方法的类型 - service
// 定义一个服务类型
service UserServiceRpc
{
    rpc Login(LoginRequest) returns(LoginResponse);
    rpc GetFriendList(GetFriendListRequest) returns(GetFriendListResponse);
}